mcp/src/main/java/io/modelcontextprotocol/client/transport/WebSocketClientTransport.java

```diff
@@ -0,0 +1,221 @@
+
+	private final Sinks.Many<Throwable> errorSink = Sinks.many().multicast().onBackpressureBuffer();
+
+	/**
+	 * The constructor for the WebSocketClientTransport.
+	 * @param uri the URI to connect to
+	 * @param clientBuilder the HttpClient builder
+	 * @param objectMapper the ObjectMapper for JSON serialization/deserialization
+	 */
+	WebSocketClientTransport(final URI uri, final HttpClient.Builder clientBuilder, final ObjectMapper objectMapper) {
+		this.uri = uri;
+		this.httpClient = clientBuilder.build();
+		this.objectMapper = objectMapper;
+	}
+
+	/**
+	 * Creates a new WebSocketClientTransport instance with the specified URI.
+	 * @param uri the URI to connect to
+	 * @return a new Builder instance
+	 */
+	public static Builder builder(final URI uri) {
+		return new Builder().uri(uri);
+	}
+
+	/**
+	 * The state of the Transport connection.
+	 */
+	public enum TransportState {
+
+		DISCONNECTED, CONNECTING, CONNECTED, CLOSED
+
+	}
+
+	/**
+	 * A builder for creating instances of WebSocketClientTransport.
+	 */
+	public static class Builder {
+
+		private URI uri;
+
+		private final HttpClient.Builder clientBuilder = HttpClient.newBuilder()
+			.version(HttpClient.Version.HTTP_1_1)
+			.connectTimeout(Duration.ofSeconds(10));
+
+		private ObjectMapper objectMapper = new ObjectMapper();
+
+		public Builder uri(final URI uri) {
+			this.uri = uri;
+			return this;
+		}
+
+		public Builder customizeClient(final Consumer<HttpClient.Builder> clientCustomizer) {
+			Assert.notNull(clientCustomizer, "clientCustomizer must not be null");
+			clientCustomizer.accept(clientBuilder);
+			return this;
+		}
+
+		public Builder objectMapper(final ObjectMapper objectMapper) {
+			this.objectMapper = objectMapper;
+			return this;
+		}
+
+		public WebSocketClientTransport build() {
+			return new WebSocketClientTransport(uri, clientBuilder, objectMapper);
+		}
+
+	}
+
+	public Mono<Void> connect(final Function<Mono<McpSchema.JSONRPCMessage>, Mono<McpSchema.JSONRPCMessage>> handler) {
+		if (!state.compareAndSet(TransportState.DISCONNECTED, TransportState.CONNECTING)) {
+			return Mono.error(new IllegalStateException("WebSocket is already connecting or connected"));
+		}
+
+		return Mono.fromFuture(httpClient.newWebSocketBuilder().buildAsync(uri, new WebSocket.Listener() {
+			private final StringBuilder messageBuffer = new StringBuilder();
+
+			@Override
+			public void onOpen(WebSocket webSocket) {
+				webSocketRef.set(webSocket);
+				state.set(TransportState.CONNECTED);
+			}
+
+			@Override
+			public CompletionStage<?> onText(WebSocket webSocket, CharSequence data, boolean last) {
+				messageBuffer.append(data);
+				if (last) {
+					final String fullMessage = messageBuffer.toString();
+					messageBuffer.setLength(0);
+					try {
+						final McpSchema.JSONRPCMessage msg = McpSchema.deserializeJsonRpcMessage(objectMapper,
+								fullMessage);
+						handler.apply(Mono.just(msg)).subscribe();
+					}
+					catch (Exception e) {
```

Comment:
I fixed this in #156. It would be nice to merge that and then we can stop catch Exception use a more fine grained type.